compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -2620,7 +2620,7 @@ where
                             if ty.is_unpin(tcx.at(DUMMY_SP), cx.param_env()) {
                                 PointerKind::UniqueBorrowed
                             } else {
-                                PointerKind::Shared
+                                PointerKind::SharedMutable
```

Comment:
Side effect on `&mut !Unpin` here -- is that part of the change just for implementation convenience, or because we want it to be legal to free the underlying object through a self-reference?

Reply:
This still changes `&mut !Unpin` to have a pointee size of `ZERO`. If we use distinct pointer kinds for `&UnsafeCell` and `&mut !Unpin`.

Being explicit with code,

```suggestion
                                if ty.is_freeze(tcx.at(DUMMY_SP), cx.param_env()) {
                                    PointerKind::SharedMutable // &mut impl !Unpin + Freeze
                                } else {
                                    PointerKind::SharedMutable // &mut impl !Unpin + !Freeze
                                }
```

Reply:
> This still changes &mut !Unpin to have a pointee size of ZERO

It already has a pointee size of 0 on master. I literally just renamed the variant.

Reply:
9131ca7dbf7c54db60943055d954b135356af189 is just the rename, but 62bf84e713be05cac8be6c52d07d7476968dac06

```diff
                    attrs.pointee_size = match kind {
-                       PointerKind::UniqueOwned => Size::ZERO,
-                       _ => pointee.size,
+                       PointerKind::UniqueBorrowed | PointerKind::Frozen => pointee.size,
+                       // If LLVM had a way of saying "dereferenceable on entry", we would use it
+                       // here, but sadly that is not currently possible.
+                       _ => Size::ZERO,
                    };
```

changes `PointerKind::SharedMutable` (previously `PointerKind::Shared`) from `pointee.size` to `Size::ZERO`.

`PointerKind::SharedMutable` (previously `PointerKind::Shared`) is used for `&impl !Freeze`, `&mut impl !Unpin + !Freeze`, *and `&mut impl !Unpin + Freeze`*.

Reply:
Oh I see, you are right -- good catch. Looks like we don"t have a test for that.